File: locuslabs-android-sdk/examples/RecommendedImplementation/app/src/main/java/com/example/locuslabs/recommendedimplementation/LocusLabsCache.java
package com.example.locuslabs.recommendedimplementation;

import android.content.Context;

import com.locuslabs.sdk.configuration.Configuration;

import java.io.File;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.TimeZone;

public class LocusLabsCache {

    private static DateFormat dateFormat = null;

    static {
        dateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH-mm-ss");
        dateFormat.setTimeZone( TimeZone.getTimeZone("UTC") );
    }

    private static final String TAG = "LocusLabsMapPack";

    public static final String LOCUSLABS = "locuslabs";

    private final File cacheDir;

    public static String getAccountId() {
        if ( null != Configuration.shared && null != Configuration.shared.getAccountId()) {
            return Configuration.shared.getAccountId();
        }
        else {
            return MainActivity.ACCOUNT_ID;
        }
    }

    public static LocusLabsCache getDefaultCache( Context context ) {
        File cache = context.getCacheDir();
        File defaultDir = new File( cache, File.separator + LOCUSLABS );
        return new LocusLabsCache( defaultDir );
    }

    protected LocusLabsCache( File directory ) {
        this.cacheDir = directory;
        if ( !this.cacheDir.exists() ) {
            this.cacheDir.mkdirs();
        }
    }

    public File getCacheDirectory() {
        return this.cacheDir;
    }

    public String pathForAsset( String asset ) {
        return this.cacheDir.getAbsolutePath() + File.separator + "assets.locuslabs.com" +  File.separator + asset;
    }

    public String venueListAsset() {
        return "accounts/" + getAccountId() + "/v4.js";
    }

    public String getLatestInstalledVersion() {
        File file = new File( this.pathForAsset( this.venueListAsset() ) );
        if ( file.exists() ) {
            return dateFormat.format( file.lastModified() );
        }
        else {
            return null;
        }
    }

}

File: locuslabs-android-sdk/examples/RecommendedImplementationKotlin/app/src/main/java/com/example/locuslabs/recommendedimplementation/LocusLabsCache.kt
package com.example.locuslabs.recommendedimplementation

import java.io.File
import java.text.DateFormat
import java.util.TimeZone
import java.text.SimpleDateFormat

import android.content.Context

import com.locuslabs.sdk.configuration.Configuration

class LocusLabsCache protected constructor(val cacheDirectory: File) {

    val venueListContents: String?
        get() = Utilities.getFileAsString(this.pathForAsset(this.venueListAsset()))

    val latestInstalledVersion: String?
        get() {
            val file = File(this.pathForAsset(this.venueListAsset()))
            return if (file.exists()) {
                dateFormat!!.format(file.lastModified())
            } else {
                null
            }
        }

    init {
        if (!this.cacheDirectory.exists()) {
            this.cacheDirectory.mkdirs()
        }
    }

    fun pathForAsset(asset: String): String {
        return this.cacheDirectory.absolutePath + File.separator + LocusLabsCache.filenameForAsset(asset)
    }

    fun venueListAsset(): String {
        return "accounts-$accountId-v3.js"
    }

    companion object {

        private var dateFormat: DateFormat? = null

        init {
            dateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH-mm-ss")
            dateFormat!!.timeZone = TimeZone.getTimeZone("UTC")
        }

        private val TAG = "LocusLabsMapPack"

        val LOCUSLABS = "locuslabs"

        fun filenameForAsset(asset: String): String {
            return asset.replace("/".toRegex(), "-")
        }

        val accountId: String
            get() = if (null != Configuration.shared && null != Configuration.shared.accountId) {
                Configuration.shared.accountId
            } else {
                MainActivity.ACCOUNT_ID
            }

        fun getDefaultCache(context: Context): LocusLabsCache {
            val cache = context.cacheDir
            val defaultDir = File(cache, File.separator + LOCUSLABS)
            return LocusLabsCache(defaultDir)
        }

        fun getCustomCache(directory: File): LocusLabsCache {
            return LocusLabsCache(directory)
        }
    }

}
